### PR TITLE
network: Route - allow to set TCP RTO

### PR DESCRIPTION
**Pull request by ssahani**
closes https://github.com/systemd/systemd/issues/28852

### CODE REVIEW

**Comment by bluca** [comment]
Can you add a test for this in `test/test-network/conf/25-route-congctl.network`?

**Review by bluca — APPROVED** [review]

File: man/systemd.network.xml [hunk 1]
```diff
@@ -1719,6 +1719,15 @@ allow my_server_t localnet_peer_t:peer recv;</programlisting>
         </listitem>
       </varlistentry>
 
+      <varlistentry>
+        <term><varname>TCPRetransmissionTimeOutSec=</varname></term>
```
Comment (yuwata) [inline man/systemd.network.xml:1723]: We usually use `Timeout` rather than `TimeOut`, e.g. `TimeoutStartSec=`

File: man/systemd.network.xml [hunk 2]
```diff
@@ -1719,6 +1719,15 @@ allow my_server_t localnet_peer_t:peer recv;</programlisting>
...
+        <listitem>
+          <para>Specifies the TCP Retransmission Time Out for the route. Takes time values in seconds.
```
Comment (yuwata) [inline man/systemd.network.xml:1725]: Please add abbreviation `(RTO)` here when it is used in later sentence.

File: src/network/networkd-route.c [hunk 3]
```diff
@@ -2752,6 +2758,58 @@ int config_parse_route_mtu(
...
+
+        r = parse_sec(rvalue, &usec);
+        if (r < 0) {
+                log_syntax(unit, LOG_WARNING, filename, line, r,
+                           "Failed to parse route TCP retransmission time out (RTO) sec '%s', ignoring: %m", rvalue);
+                return 0;
+        }
+
+        if (usec != USEC_INFINITY &&
+            DIV_ROUND_UP(usec, USEC_PER_SEC) > UINT32_MAX) {
```
Comment (yuwata) [inline src/network/networkd-route.c:2801]: `USEC_PER_MSEC`

File: src/network/networkd-route.c [hunk 4]
```diff
@@ -2752,6 +2758,58 @@ int config_parse_route_mtu(
...
+                log_syntax(unit, LOG_WARNING, filename, line, 0,
+                           "Route TCP retransmission time out (RTO) = must be in the range 0...%"PRIu32"ms, ignoring: %s", UINT32_MAX, rvalue);
```
Comment (yuwata) [inline src/network/networkd-route.c:2803]: - `(RTO) = must be` -> `(RTO) must be`
- `...` -> `…`

File: src/network/networkd-route.c [hunk 5]
```diff
@@ -1247,6 +1247,12 @@ static int route_configure(const Route *route, uint32_t lifetime_sec, Link *link
                         return r;
         }
 
+        if (route->tcp_rto_usec > 0) {
+                r = sd_netlink_message_append_u32(m, RTAX_RTO_MIN, route->tcp_rto_usec / USEC_PER_MSEC);
```
Comment (yuwata) [inline src/network/networkd-route.c:1251]: `DIV_ROUND_UP()` should be used.
Also, `USEC_INFINITY` should be handled separately.

File: src/network/networkd-route.c [hunk 6]
```diff
@@ -2752,6 +2758,58 @@ int config_parse_route_mtu(
...
+        }
+
+        r = parse_sec(rvalue, &usec);
+        if (r < 0) {
+                log_syntax(unit, LOG_WARNING, filename, line, r,
+                           "Failed to parse route TCP retransmission time out (RTO) sec '%s', ignoring: %m", rvalue);
+                return 0;
+        }
+
+        if (usec != USEC_INFINITY &&
```
Comment (yuwata) [inline src/network/networkd-route.c:2800]: Hmm... The kernel handle UINT32_MAX as INFINITY or so?? If not, Please refuse infinity.

**Review by yuwata — COMMENTED** [review]
Sorry for late review.

**Pull request merged**